[receiver/k8sobjects] Support running in kubernetes leader election mode. (#24662)
open-telemetry/opentelemetry-collector-contrib: receiver/k8sobjectsreceiver/README.md

        mode: watch
        group: events.k8s.io
        namespaces: [default]

  k8sobjects/leader-election:
    leader_election:
      enabled: true
      lock_name: "k8sobjects-leader-election"
    auth_type: serviceAccount
    objects:
      - name: pods
        mode: pull
        label_selector: environment in (production),tier in (frontend)
        field_selector: status.phase=Running
        interval: 15m
      - name: events
        mode: watch
        group: events.k8s.io
        namespaces: [default]
```

Brief description of configuration properties:

> - `leader_election`: running in Kubernetes with `leader election mode`, this means that multiple instances are running, but only one is active at a time, and if it fails, another one is elected as leader and takes its place.

Reviewer: I think we should mention the additional access needed by the collector's service account to perform leader election here, and also update the helm chart.

I believe this snippet should cover rbac for leaderelection.
```
- apiGroups:
  - coordination.k8s.io
  resources:
  - leases
  verbs:
  - create
  - get
  - update
  ```
  
  In case users prefer to make this stricter and are providing their own election ids, they can update the role to something like
```

- apiGroups:
  - coordination.k8s.io
  resources:
  - leases
  verbs:
  - create
- apiGroups:
  - coordination.k8s.io
  resourceNames:
  - <ELECTION ID>
  resources:
  - leases
  verbs:
  - get
  - update
  ```

Author: > and also update the helm chart.

yes, I will update soon in [opentelemetry-helm-charts](https://github.com/open-telemetry/opentelemetry-helm-charts) repo after this PR get merged. :-P

  - `enabled` (default = `false`): whether run in leader election mode.
  - `lock_name` (default = `k8sobjects`): the identity name of holder, will use component's ID if not set.
  - `lease_duration` (default = `15s`): the duration that non-leader candidates will wait to force acquire leadership.

>   - `re_new_deadline` (default = `10s`): the duration that the acting master will retry refreshing leadership before giving up.

Reviewer: nit: this should be `renew_deadline`

    auth_type: serviceAccount
    objects:
      - name: pods
        mode: pull
        label_selector: environment in (production),tier in (frontend)
        field_selector: status.phase=Running
        interval: 15m
      - name: events
        mode: watch
        group: events.k8s.io
        namespaces: [default]
```

Brief description of configuration properties:
- `leader_election`: Leader election mode helps to ensure that only one instance (as leader instance) is collecting `k8sobjects` data, and if the leader instance fails, another one is elected as leader and takes its place. 
*Note*: You need to add additional RBAC to perform the leader election mode. See below [Leader Election RBAC](#RBAC-Leader-Election) for more information.
  - `enabled` (default = `false`): whether run in leader election mode.
  - `lock_name` (default = `k8sobjects`): the identity name of holder, will use component's ID if not set.
  - `lease_duration` (default = `15s`): the duration that non-leader candidates will wait to force acquire leadership.
  - `renew_deadline` (default = `10s`): the duration that the acting master will retry refreshing leadership before giving up.

>   - `retry_period` (default = `2s`): the duration the LeaderElector clients should wait between tries of actions. 

Reviewer 2: I'm not sure what this means just by reading this doc. Are "LeaderElector clients" referred to as "non-leader candidates" in a previous item?

        mode: watch
        group: events.k8s.io
        namespaces: [default]

  k8sobjects/leader-election:
    leader_election:
      enabled: true
      lock_name: "k8sobjects-leader-election"
    auth_type: serviceAccount
    objects:
      - name: pods
        mode: pull
        label_selector: environment in (production),tier in (frontend)
        field_selector: status.phase=Running
        interval: 15m
      - name: events
        mode: watch
        group: events.k8s.io
        namespaces: [default]
```

Brief description of configuration properties:

> - `leader_election`: running in Kubernetes with `leader election mode`, this means that multiple instances are running, 

Reviewer 2: What happens if I have two instances of this component for the same collector? Is only one of them going to be elected leader? Why would I use this? Should I always use this, or only on specific situations?

Author: I updated the docs.

        mode: watch
        group: events.k8s.io
        namespaces: [default]

  k8sobjects/leader-election:
    leader_election:
      enabled: true
      lock_name: "k8sobjects-leader-election"
    auth_type: serviceAccount
    objects:
      - name: pods
        mode: pull
        label_selector: environment in (production),tier in (frontend)
        field_selector: status.phase=Running
        interval: 15m
      - name: events
        mode: watch
        group: events.k8s.io
        namespaces: [default]
```

Brief description of configuration properties:

> - `leader_election`: We usually run multiple instances to ensure high availability of `otel-col`, but the `k8sobjectsreceiver` collected data will be duplicated, 
> so we can use the `leader election mode` to ensure that only one instance(as leader instance) is collecting `k8sobjects` data, and the other instances are the alternative nodes.
> and if the leader instance fails, another one is elected as leader and takes its place. 
> *Note*: We need to add additional permissions in RBAC to perform `leader election mode`, See below [Leader Election RBAC](#RBAC-Leader-Election) for more information.

Reviewer 3: ```suggestion
- `leader_election`: Leader election mode helps to ensure that only one instance (as leader instance) is collecting `k8sobjects` data, and if the leader instance fails, another one is elected as leader and takes its place. 
*Note*: You need to add additional RBAC to perform the leader election mode. See below [Leader Election RBAC](#RBAC-Leader-Election) for more information.
```
